### PR TITLE
[inductor] optimize isa dry compile time.

### LINKED ISSUE #100378: VecISA.__bool__ is very expensive (nearly a second) on startup
### 🐛 Describe the bug

VecISA currently requires compiling a cpp program to test for capabilities, similar to `./configure`. We do this compilation **on every new process invocation**, dissimilar to `./configure`. After performing the capability query, we should cache it and reuse the cache entry. This will save us about 1s cold start time on CPU inductor. 

### Versions

master

cc @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @wconstab @soumith @ngimel

### PR DESCRIPTION
**Pull request by xuhancn**
Fixes #100378
Original issue caused by startup dry compile need cost almost 1 second.

This PR add compiler version info, isa build options and pytorch version info to the test binary path hash.
So same compile, same isa and same pytorch can skip the dry compile.

Local test:
First time:
image
We need to compile all c++ modules and it cost 16.5s.

Second time:
image
We skipped dry compile due to the same isa fingerprint. It is only cost 0.36s.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -1050,6 +1052,40 @@ def is_clang() -> bool:
...
+    # Mutiple line to one line string.
+    version_string = version_string.replace("\r", "_")
+    version_string = version_string.replace("\n", "_")
+    return version_string
+
+
+def _get_isa_dry_compile_fingerprint(isa_flags: str) -> str:
+    # ISA dry compile will cost about 1 sec time each startup time.
+    # Please check the issue: https://github.com/pytorch/pytorch/issues/100378
+    # Acturaly, dry compile is checking compile capability for ISA.
```
Comment (jgong5) [inline torch/_inductor/codecache.py:1079]: nit
```suggestion
    # Actually, dry compile is checking compile capability for ISA.
```

**Review by jgong5 — APPROVED** [review]

**Comment by ezyang** [comment]
I guess we can land this as is, but you're still doing multiple(!) subprocess calls, and I would hope that there was some way to architect this where those were not needed either.

**Comment by xuhancn** [comment]
> I guess we can land this as is, but you're still doing multiple(!) subprocess calls, and I would hope that there was some way to architect this where those were not needed either.

Here has some `lru_cache` to cache result. What optimization you still want to do?

**Comment by xuhancn** [comment]
https://github.com/pytorch/pytorch/pull/124602/commits/3e0915654b9618fd77d77dbd41bd73d6fce3fada @ezyang I add a cache to `pick_vec_isa` and it would reduce the callee function run.

**Comment by xuhancn** [comment]
@malfet Any comment to me for next step?

**Comment by ezyang** [comment]
lru_cache doesn't help, you need a cache that persists across processes.

**Comment by xuhancn** [comment]
> lru_cache doesn't help, you need a cache that persists across processes.

I guess you means mutiple process call `pick_vec_isa` to get isa for cpp build. 
I think what I can do is that write the isa info to the temp file. But still a problem, we always need to get compiler version information to ensure the compiler's capability can cover the isa level. At the same time, get compile version still need to call `subprocess`. So I think we maybe not have potential to reduce call `subprocess`.

**Comment by xuhancn** [comment]
@ezyang we can land this PR firstly.
If any other optimize idea let create new PR to land.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
ci fails look real

**Comment by xuhancn** [comment]
> ci fails look real

Let me debug for it.

**Comment by xuhancn** [comment]
> ci fails look real

@ezyang the CI was failed as the reason that `lru_cache` for `pick_vec_isa` cache the value not match the test case. After revert the related code, it works well.

**Pull request closed without merging**